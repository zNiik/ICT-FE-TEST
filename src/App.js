import React, { useCallback, useState } from 'react'
import classNames from 'classnames/bind';
import styles from './app.css'
import useCountdowns from './components/Countdowns/hooks/useCountdowns';
import Countdowns from './components/Countdowns';
import Select from './components/Select'

let cx = classNames.bind(styles);

const App = () => {
  const [catId, setCatId] = useState();
  const countdowns = useCountdowns();
  const { categories, loading } = countdowns;
  
  const onSelect = useCallback((catId) => {
    setCatId(catId);
  }, []);
  
  return (
    <React.Fragment>
      <div className={styles.header}>
        <div className={styles.container}>
          <h1>ICT Test</h1>
          <Select options={categories} loading={loading} onSelect={onSelect}  />
        </div>
      </div>

      <div className={cx({ wrapper: true, 'wrapper--is-loading': loading })}>
        <div className={styles.container}>
          <Countdowns {...countdowns} catIdFilter={catId} />
        </div>
      </div>
    </React.Fragment>
  )
}

export default App