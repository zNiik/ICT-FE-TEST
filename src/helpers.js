import ict from './assets/ict.svg'
import beer from './assets/beer.jpg'
import sleep from './assets/sleep.jpg'
import tomorrow from './assets/tomorrow.jpg'
import yesterday from './assets/yesterday.jpg'
import yoga from './assets/yoga.jpg'

const Images = { ict, beer, sleep, tomorrow, yesterday, yoga }

export const Category = {
  WORK: 'Work',
  HEALTH: 'Health',
  LEISURE: 'Leisure'
}

export const categories = [
  { id: 1, name: Category.WORK },
  { id: 2, name: Category.HEALTH },
  { id: 3, name: Category.LEISURE }
]

const M_TO_S = 60
const H_TO_S = M_TO_S * 60
const D_TO_S = H_TO_S * 24

export const getDateString = (seconds, minutes = 0, hours = 0, days = 0, sign = 1) => {
  const totalSeconds = sign * (seconds + minutes * M_TO_S + hours * H_TO_S + days * D_TO_S)
  const now = new Date()
  now.setSeconds(now.getSeconds() + totalSeconds)
  return now.toUTCString()
}

export const countdowns = [
  { id: 1, categoryId: 2, title: "Time to sleep", image: 'sleep', startsAt: getDateString(10, 8)},
  { id: 2, categoryId: 2, title: "Yoga Time", image: 'yoga', startsAt: getDateString(10)},
  { id: 3, categoryId: 3, title: "Beer o'clock", image: 'beer', startsAt: getDateString(4, 3)},
  { id: 4, categoryId: 1, title: "Meeting #1", image: 'yesterday', startsAt: getDateString(4, 3, 2, 1, -1)},
  { id: 5, categoryId: 1, title: "Meeting #2", image: 'tomorrow', startsAt: getDateString(4, 3, 2, 1)}
]

export const getImageSrc = (name) => Images[name];

export const fetchCategories = () => 
  new Promise(resolve => {
    setTimeout(() => resolve(categories), 600)
  })

export const fetchCountdowns = () => 
  new Promise(resolve => {
    setTimeout(() => resolve(countdowns), 1000)
  })

