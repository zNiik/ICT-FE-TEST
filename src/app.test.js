import React from 'react';
import { mount } from 'enzyme';
import { waitFor } from '@testing-library/react';
import * as helpers from './helpers'
import App from './App';

jest.useFakeTimers();

describe('<App /> loading state', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(<App />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it ('should initially render loading state for <Select /> and <Countdowns /> component', () => {
    expect(wrapper.find('Select Loader')).toHaveLength(1);
    expect(wrapper.find('Countdowns Loader')).toHaveLength(1);
  });
});

describe('<App /> loaded', () => {
  let wrapper;
  
  beforeEach(async () => {
    wrapper = mount(<App />);

    await waitFor(async () => {
      await new Promise((resolve) => setTimeout(resolve, 2000));
    });
  });

  it('should renders <Select /> and <Countdowns /> with proper data', async () => {
    await waitFor(async () => {
      wrapper.update();
    });
    
    expect(wrapper.find('select').children()).toHaveLength(helpers.categories.length + 1);
    expect(wrapper.find('[className="countdowns-list"]').children()).toHaveLength(helpers.countdowns.length);
  });

  it('should filter by the selected category', async () => {
    await waitFor(async () => {
      wrapper.update();
    });

    wrapper.find('select').simulate('change', { target: { value: '3' }});

    expect(wrapper.find('[className="countdowns-list"]').children()).toHaveLength(1);
  });
});

