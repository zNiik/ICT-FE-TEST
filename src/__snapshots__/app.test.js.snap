// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> loading state renders correctly 1`] = `ReactWrapper {}`;
